### PR TITLE
blog: integrate OIDC with oVirt using Keycloak and LDAP user federation

### PR DESCRIPTION
**Pull request by noriravi**
blog: integrate OIDC with oVirt using Keycloak and LDAP user federation

I confirm that this pull request was submitted according to the [contribution guidelines](https://github.com/oVirt/ovirt-site/blob/master/CONTRIBUTING.md): (please @mention yourself to sign)

### CODE REVIEW

File: source/blog/20019-01-22-sso-keycloak-ovirt-openidc-389ds.html.md [hunk 1]
```diff
@@ -0,0 +1,267 @@
+---
+title: oVirt OIDC with keycloak using 389ds user federation
```
Comment (mwperina) [inline source/blog/20019-01-22-sso-keycloak-ovirt-openidc-389ds.html.md]: I'd simplify it to `Federate oVirt autthetication to OpenID Connect infrastructure`

File: source/blog/20019-01-22-sso-keycloak-ovirt-openidc-389ds.html.md [hunk 2]
```diff
@@ -0,0 +1,267 @@
...
+As OIDC is not integrated into oVirt directly, we use Apache to do the OIDC authentication for us. The [mod_auth_openidc](https://github.com/zmartzone/mod_auth_openidc) module nicely covers all needed functionality.
+
+Overview
+
+Integrate with external OpenID Connect Identity Provider (IDP) to provide Single Sign-On  (SSO) across products that use the IDP for authenticating users. We currently have oVirt SSO for providing unified authentication across Administrator and VM portals. The oVirt SSO also provides tokens for REST API clients and supports bearer authentication to reuse the tokens to access engine api. With external IDP integration the internal oVirt SSO is disabled and browser users will be redirected to the external IDP for authentication. Once authenticated the users can access both Admin and VM portal as they normally do. REST API clients don't have to change, they can still obtain a token from engine SSO and use the token for bearer authentication to access the engine’s REST API. Engine SSO acts as a proxy obtaining the token from external IDP on behalf of the rest api client.
```
Comment (mwperina) [inline source/blog/20019-01-22-sso-keycloak-ovirt-openidc-389ds.html.md]: "the tokens" -> "tokens"

Comment (mwperina) [reply source/blog/20019-01-22-sso-keycloak-ovirt-openidc-389ds.html.md]: "engine api" -> "oVirt engine RESTAPI"

Comment (mwperina) [reply source/blog/20019-01-22-sso-keycloak-ovirt-openidc-389ds.html.md]: "Once authenticated the users" -> "After successful authentication users"

Comment (mwperina) [reply source/blog/20019-01-22-sso-keycloak-ovirt-openidc-389ds.html.md]: "Admin and VM portal" -> "Admin and VM portals"

Comment (mwperina) [reply source/blog/20019-01-22-sso-keycloak-ovirt-openidc-389ds.html.md]: "engine’s REST API" -> "oVirt engine RESTAPI"

Comment (mwperina) [reply source/blog/20019-01-22-sso-keycloak-ovirt-openidc-389ds.html.md]: "rest api client" -> "RESTAPI client"

File: source/blog/20019-01-22-sso-keycloak-ovirt-openidc-389ds.html.md [hunk 3]
```diff
@@ -0,0 +1,267 @@
...
+
+To access Administrator portal the users need to belong to LDAP group named “ovirt-administrator” or an LDAP group that has been manually granted permissions to access Admin portal.
```
Comment (mwperina) [inline source/blog/20019-01-22-sso-keycloak-ovirt-openidc-389ds.html.md]: "the users" -> "users"

File: source/blog/20019-01-22-sso-keycloak-ovirt-openidc-389ds.html.md [hunk 4]
```diff
@@ -0,0 +1,267 @@
...
+![](../images/blog/2019-01-22/user_search.png)
+
+Now that user federation is working and is integrated, let's configure the OIDC client. 
+
+![](../images/blog/2019-01-22/add_client.png)
+
+**IMPORTANT** After the client is created, we must change `Access Token Lifespan` to `10 days`. This is required because oVirt manages the token lifespan and we do not want Keycloak invalidating the tokens.
+![](../images/blog/2019-01-22/life_span.png)
+
+Make a note of client id and client secret from the cleint credentials page, the client id and secret is needed to configure the apache mod_auth_openidc and also ovirt-engine.
```
Comment (mwperina) [inline source/blog/20019-01-22-sso-keycloak-ovirt-openidc-389ds.html.md]: "cleint" -> "client"

Comment (mwperina) [reply source/blog/20019-01-22-sso-keycloak-ovirt-openidc-389ds.html.md]: "apache" -> "Apache"

File: source/blog/20019-01-22-sso-keycloak-ovirt-openidc-389ds.html.md [hunk 5]
```diff
@@ -0,0 +1,267 @@
...
+Create the scopes need by ovirt engine, ovirt-app-api and ovirt-app-admin
+![](../images/blog/2019-01-22/scope_ovirt_app_api.png)
+
+![](../images/blog/2019-01-22/scope_ovirt_app_admin.png)
+
+Set up optional client scopes for the ovirt-engine client
+![](../images/blog/2019-01-22/client_scopes.png)
+
+## mod_auth_openidc configuration
+First we need to configure oVirt's apache. SSH to the oVirt engine and install mod_auth_openidc and ovirt-engine-extension-aaa-misc.
```
Comment (mwperina) [inline source/blog/20019-01-22-sso-keycloak-ovirt-openidc-389ds.html.md]: "First we need to configure oVirt's apache" -> "At first we need to configure Apache on oVirt engine machine"

File: source/blog/20019-01-22-sso-keycloak-ovirt-openidc-389ds.html.md [hunk 6]
```diff
@@ -0,0 +1,267 @@
...
+yum install -y ovirt-engine-extension-aaa-misc
+```
+
+We need to create a file `/etc/httpd/conf.d/ovirt-openidc.conf` to configure mod_auth_openidc as follows:
+
+```
+LoadModule auth_openidc_module modules/mod_auth_openidc.so
+ServerName engineopenid.example.com
+
+OIDCProviderMetadataURL https://keycloak.example.com:8443/auth/realms/master/.well-known/openid-configuration
```
Comment (mwperina) [inline source/blog/20019-01-22-sso-keycloak-ovirt-openidc-389ds.html.md]: Please mention somewhere at beginning that users need to replace "keycloak.example.com"  and all other example.com URLs with real URLs

File: source/blog/20019-01-22-sso-keycloak-ovirt-openidc-389ds.html.md [hunk 7]
```diff
@@ -246,3 +248,20 @@ connection.close()
 
 Login to webadmin, user1 belongs to group ovirt-administrator in ldap, so should get access webadmin automatically.
```
Comment (mwperina) [inline source/blog/20019-01-22-sso-keycloak-ovirt-openidc-389ds.html.md]: "so should get access webadmin automatically" -> "so it should be able to login to webadmin without any additional permission settings"

**Comment by mwperina** [comment]
@machacekondra @gregsheremeta

**Comment by mwperina** [comment]
@dominikholler

File: source/blog/20019-01-22-sso-keycloak-ovirt-openidc-389ds.html.md (hunk 5 above)
Comment (gregsheremeta) [reply source/blog/20019-01-22-sso-keycloak-ovirt-openidc-389ds.html.md]: "At first" isn't a typical phrase, actually. Ravi's version is more correct :)

**Comment by gregsheremeta** [comment]
@noriravi the build is failing on bad date (20019)

rn 20019-01-22-sso-keycloak-ovirt-openidc-389ds.html.md  -> 2019-01-22-sso-keycloak-ovirt-openidc-389ds.html.md

File: source/blog/20019-01-22-sso-keycloak-ovirt-openidc-389ds.html.md (hunk 5 above)
Comment (mwperina) [reply source/blog/20019-01-22-sso-keycloak-ovirt-openidc-389ds.html.md]: OK :-)

File: source/blog/2019-01-22-sso-keycloak-ovirt-openidc-389ds.html.md [hunk 8]
```diff
@@ -0,0 +1,271 @@
...
+
+To access Administrator portal users need to belong to LDAP group named “ovirt-administrator” or an LDAP group that has been manually granted permissions to access Admin portal.
+
+Note
+
+In the examples and configuration below we user keycloak.example.com as the Keycloak server URL and engineopenid.example.com as the oVirt engine server URL. Please note that these should be replaced with real URLs from your setup.
```
Comment (machacekondra) [inline source/blog/2019-01-22-sso-keycloak-ovirt-openidc-389ds.html.md]: s/user/use

Comment (machacekondra) [reply source/blog/2019-01-22-sso-keycloak-ovirt-openidc-389ds.html.md]: Maybe quote the the URL, and actually those in the example are fqdn not URLs

**Comment by gregsheremeta** [comment]
[ squashed and rebased ]

**Comment by mwperina** [comment]
Greg, could you please merge?
@gregsheremeta

**Pull request merged**